crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -81,200 +46,178 @@ fn load_spec(bundle: String) -> Result<oci::Spec, Error> {
     Ok(spec)
 }
 
-pub fn prepare_module(
-    engine: wasmtime::Engine,
-    spec: &oci::Spec,
-    stdin_path: String,
-    stdout_path: String,
-    stderr_path: String,
-) -> Result<(WasiCtx, Module, String), WasmtimeError> {
-    debug!("opening rootfs");
-    let rootfs = oci_wasmtime::get_rootfs(spec)?;
-    let args = oci::get_args(spec);
-    let env = oci_wasmtime::env_to_wasi(spec);
-
-    debug!("setting up wasi");
-    let mut wasi_builder = WasiCtxBuilder::new()
-        .args(args)?
-        .envs(env.as_slice())?
-        .preopened_dir(rootfs, "/")?;
-
-    debug!("opening stdin");
-    let stdin = maybe_open_stdio(&stdin_path).context("could not open stdin")?;
-    if let Some(sin) = stdin {
-        wasi_builder = wasi_builder.stdin(Box::new(sin));
-    }
-
-    debug!("opening stdout");
-    let stdout = maybe_open_stdio(&stdout_path).context("could not open stdout")?;
-    if let Some(sout) = stdout {
-        wasi_builder = wasi_builder.stdout(Box::new(sout));
-    }
-
-    debug!("opening stderr");
-    let stderr = maybe_open_stdio(&stderr_path).context("could not open stderr")?;
-    if let Some(serr) = stderr {
-        wasi_builder = wasi_builder.stderr(Box::new(serr));
-    }
-
-    debug!("building wasi context");
-    let wctx = wasi_builder.build();
-    debug!("wasi context ready");
-
-    let start = args[0].clone();
-    let mut iterator = start.split('#');
-    let mut cmd = iterator.next().unwrap().to_string();
-
-    let stripped = cmd.strip_prefix(std::path::MAIN_SEPARATOR);
-    if let Some(strpd) = stripped {
-        cmd = strpd.to_string();
-    }
-    let method = iterator.next().unwrap_or("_start");
-
-    let mod_path = oci::get_root(spec).join(cmd);
-    debug!("loading module from file");
-    let module = Module::from_file(&engine, mod_path)
-        .map_err(|err| Error::Others(format!("could not load module from file: {}", err)))?;
-
-    Ok((wctx, module, method.to_string()))
-}
-
 impl Instance for Wasi {
     type E = wasmtime::Engine;
-    fn new(_id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self {
-        let cfg = cfg.unwrap(); // TODO: handle error
+    fn new(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self {
+        // TODO: there are failure cases e.x. parsing cfg, loading spec, etc.
+        // thus should make `new` return `Result<Self, Error>` instead of `Self`
```

Comment:
This was discussed before: https://github.com/containerd/runwasi/pull/54#issuecomment-1403269766

IMHO we should aim to remove as many `unwrap()` calls as possible from shim "main thread", because a library should not panic easily.

Reply:
We can raise an issue for tracking this one. No need to change the API in this PR.

Reply:
Created one and resolving this convo.